src/lua/protobuf.lua

```diff
@@ -0,0 +1,723 @@
+-- input value.
+local function check_integer(field_def, value)
+    if (type(value) == 'number' and math.ceil(value) ~= value) then
+        error(('Input number value %f for %q is not integer'):format(
+            value, field_def.name))
+    elseif type(value) == 'cdata' and not is_number64(value) then
+        error(('Input cdata value %q for %q field is not integer'):format(
+            ffi.typeof(value), field_def.name))
+    end
+end
+
+-- }}} Global helpers
+
+
+-- {{{ is_scalar, is_enum, is_message
+
+local function is_scalar(field_def)
+    return scalars[field_def.type]
+end
+
+local function is_enum(protocol, field_def)
+    return protocol[field_def.type].type == 'enum'
+end
+
+local function is_message(protocol, field_def)
+    return protocol[field_def.type].type == 'message'
+end
+
+-- }}} is_scalar, is_enum, is_message
+
+
+-- {{{ Validations
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function validate_table_is_array(field_def, data)
+    assert(type(data) == 'table')
+    local key_count = 0
+    local min_key = 1/0
+    local max_key = -1/0
+    for k, _ in pairs(data) do
+        if type(k) ~= 'number' then
+            error(('Input array for %q repeated field ' ..
+                'contains non-numeric key: %q'):format(field_def.name, k))
+        end
+        if k - math.floor(k) ~= 0 then
+            error(('Input array for %q repeated field contains ' ..
+                'non-integer numeric key: %q'):format(field_def.name, k))
+        end
+        key_count = key_count + 1
+        min_key = math.min(min_key, k)
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(field_def.name,
+            field_def.type, type(value)))
```

Comment:
As for me, 'unsupported or colliding' doesn't add much to the problem explanation. I would even say that it raises more questions. For example, what types are 'unsupported' and what types are 'colliding'? What does 'colliding' means in this context? Conflicting?

I would drop this part of the error message, because it seems obvious from the existing message why is the problem: a value of an improper type is passed.